### PR TITLE
Fix multiple evaluations of mutable QNode

### PR DESCRIPTION
**Pull request by antalszava**
**Context:**
Currently, a mutable `QNode` evaluated multiple times fails if there is a change in the list `Variable` instances that are being used. This can be seen currently when running the QAOA Maxcut tutorial.

**Description of the Change:**
Adds a slightly more sophisticated check for when to generate the list of `Variable` instance in the `BaseQNode._construct` method.

**Benefits:**
A very lightweight check is being used for mutable `QNode`. This assumes that the input `args` argument can be converted to a `numpy.ndarray` and that the `shape` attribute of this `numpy.ndarray` will uniquely identify the arguments passed to `_construct`. The use of methods from the `inspect` library is deliberately avoided such that no considerable overhead is introduced.

The test suite passes including a new Integration test getting the edge case similar to the one in the QAOA MaxCut tutorial.

**Possible Drawbacks:**
This PR addresses the case of when the `QNode` is mutated in its args in a specific variety of ways (querying the shape of a `numpy.ndarray`).

Further TODO items are contained in the following issues:
#506 (this might also contain an unexpected behaviour when using keyword arguments)
#507  

**Related GitHub Issues:**
https://github.com/XanaduAI/qml/issues/36

### CODE REVIEW

**Comment by smite** [comment]
> This assumes that the input args argument can be converted to a numpy.ndarray and that the shape attribute of this numpy.ndarray will uniquely identify the arguments passed to _construct.

I don't think you can assume this. Currently each of the qfunc input args is allowed to be an arbitrary nested structure of `Iterables`, basically anything that `_flatten` can eat. You can always convert such an object into an object ndarray, but in general the shape of the nested structure is not reflected in the shape of the array (unless it *happens* to exactly correspond to an array).

File: tests/test_default_qubit.py [hunk 1]
```diff
@@ -1094,6 +1094,38 @@ def circuit():
...
+        weights = np.array([[phi], [theta]])
+        res = circuit(weights, n_layers=2)
+        exp = np.cos(phi+theta)
+        assert np.allclose(res, exp, atol=tol, rtol=0)
+
+        weights = np.array([[phi], [theta], [varphi]])
+        res = circuit(weights, n_layers=3)
+        exp = np.cos(phi+theta+varphi)
+        assert np.allclose(res, exp, atol=tol, rtol=0)
+
```
Comment (smite) [inline tests/test_default_qubit.py]: The simplest test case I could think of that reveals the bug is

```
import pennylane as qml

dev = qml.device("default.qubit", wires=1)

@qml.qnode(dev)
def node(x, n=1):
    for k in range(n):
        qml.RX(x[k], wires=0)
    return qml.expval(qml.PauliZ(wires=0))

node([0.1], n=1)
node([0.1, 0.2], n=2)  # bug: raises an IndexError
```

Comment (antalszava) [reply tests/test_default_qubit.py]: Sure, thanks for that! I'll simplify the way `weights` is handled such that the test is better readable

**Comment by antalszava** [comment]
Thanks for the comments @smite! One thing to note about this PR is that it is not meant to be extensive, but rather a fix to the existing tests and tutorials that we have. The edge case that you've identified could be picked up in subsequent PRs that addresses these issues, where the creation and caching of `Variable` instances inside the `BaseQNode._construct` is done in a more sophisticated way.

**Comment by smite** [comment]
I'm not sure it's an edge case, at some point when the Variable system was being designed these arbitrarily nested iterables were a specifically requested feature.

But I assume all the current tutorials and examples run with the new check?

**Comment by antalszava** [comment]
Sure! I'm not too familiar with the exact requirements that were set back then, but it might very well be that the type of mutable `QNode` that you have included was not regarded in the design of the logic. 

Yes, the scope of this PR would be to:
* Add a test that reproduces that mutable `QNode` setting that fails for the QAOA MaxCut tutorial
* Alter the logic of the creation of `Variable` instances in PennyLane such that the new test and the existing test suite passes

Currently, this PR doesn't want to achieve more than that, as a further change in the logic would require more thought on how to perform these checks in an efficient way.

**Comment by antalszava** [comment]
Sure! I'm not too familiar with the exact requirements that were set back then, but it might very well be that the type of mutable `QNode` that you have included was not regarded in the design of the logic or that logic was also affected by the previous circuit drawer PR and there were no explicit tests for them.

Yes, the scope of this PR would be to:
* Add a test that reproduces that mutable `QNode` setting that fails for the QAOA MaxCut tutorial
* Alter the logic of the creation of `Variable` instances in PennyLane such that the new test and the existing test suite passes

Having these changes, the test suite passes and the QAOA MaxCut tutorial (the one that motivated this PR) passes.

Currently, this PR doesn't want to achieve more than that, as a further change in the logic would require more thought on how to perform these checks in an efficient way.

**Comment by antalszava** [comment]
The drop in coverage comes from the following lines:
pennylane/qnodes/base.py                                264      3    99%   354-357

These are parts of `_determine_structured_variable_name` method related to the string representation of `Variable` instances.

File: pennylane/qnodes/base.py [hunk 2]
```diff
@@ -474,8 +474,12 @@ def _construct(self, args, kwargs):
         """
         # pylint: disable=attribute-defined-outside-init, too-many-branches, too-many-statements
 
-        if self.arg_vars is None or self.kwarg_vars is None:
```
Comment (johannesjmeyer) [inline pennylane/qnodes/base.py]: So basically, just remove this line?

Comment (antalszava) [reply pennylane/qnodes/base.py]: Yes, that's very true! The reason for reverting the logic is more on the side of performance, see the comment at https://github.com/XanaduAI/pennylane/pull/446/files#r370309305

Comment (josh146) [reply pennylane/qnodes/base.py]: Yep --- @antalszava did some benchmarking, and found that removing the caching (and having `_make_variables` be called with every evaluation) adds a _lot_ of overhead, mostly due to the use of `functools.inspect`.

Comment (josh146) [reply pennylane/qnodes/base.py]: (in one particular case, a QNode evaluation loop was 12s slower with this line removed, I believe?)

File: pennylane/variable.py [hunk 3]
```diff
@@ -156,27 +156,31 @@ def val(self):
...
                 not self.is_kwarg
                 and VariableRef.positional_arg_values is not None
                 and len(VariableRef.positional_arg_values) > self.idx
             ):
                 return str(round(self.val, 3))
 
-        if self.mult != 1:
-            return "{}*{}".format(str(round(self.mult, 3)), self.name)
+        if :
+        '''
```
Comment (josh146) [inline pennylane/variable.py]: I would make this a multiline comment, rather than a multiline string statement (it also allows you to remove the pylint check above)

Comment (josh146) [reply pennylane/variable.py]: Also, how come the string above doesn't match the previous code? Ideally, we should leave as much of the previous code as-is, just commented out (and with instructions for what to do when the circuit drawer functionality is returned).

File: pennylane/variable.py [hunk 4]
```diff
@@ -156,27 +156,31 @@ def val(self):
...
-    def render(self, show_name_only=False):
+    def render(self):
         """Returns a string representation of the VariableRef.
 
-        Args:
-            show_name_only (bool, optional): Render the name instead of the value.
-
         Returns:
             str: A string representation of the VariableRef
         """
```
Comment (josh146) [inline pennylane/variable.py]: A question --- would it be cleaner to leave this method as is? Since the qnode._construct logic has been reverted, the `show_name_only=True` case will never occur here, and it means less changes when we want to return this functionality.

File: pennylane/circuit_drawer/representation_resolver.py [hunk 5]
```diff
@@ -29,9 +29,12 @@ class RepresentationResolver:
         show_variable_names (bool, optional): Show variable names instead of variable values.
     """
 
-    def __init__(self, charset=UnicodeCharSet, show_variable_names=False):
```
Comment (josh146) [inline pennylane/circuit_drawer/representation_resolver.py]: Does this need to be removed here? That is, is it sufficient to leave this keyword argument, as long as it is never called in _construct?

File: tests/circuit_drawer/test_circuit_drawer.py [hunk 6]
```diff
@@ -641,6 +644,22 @@ def qfunc(a, w):
...
+        @qml.qnode(dev)
+        def qfunc(a, w):
+            qml.Hadamard(0)
+            qml.CRX(a, wires=[0, 1])
+            qml.Rot(w[0], w[1], w[2], wires=[1])
+            qml.CRX(-a, wires=[0, 1])
+
+            return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))
+
+        res = qfunc(2.3, [1.2, 3.2, 0.7])
```
Comment (josh146) [inline tests/circuit_drawer/test_circuit_drawer.py]: :+1:

File: tests/qnodes/test_qnode_base.py [hunk 7]
```diff
@@ -830,6 +830,54 @@ def non_mutable_circuit(x, *, c=None):
...
+                    qml.RX(x[k][j], wires=0)
+            return qml.expval(qml.PauliZ(wires=0))
+
+        res = node([[0.1], [0.2]], n=1)
+        exp = np.cos(sum([0.1] + [0.2]))
+        assert np.allclose(res, exp, atol=tol, rtol=0)
+
+        res = node([[0.1], [0.2, 0.3]], n=2)
+        exp = np.cos(sum([0.1] + [0.2, 0.3]))
+        assert np.allclose(res, exp, atol=tol, rtol=0)
```
Comment (josh146) [inline tests/qnodes/test_qnode_base.py]: Nice!

File: pennylane/qnodes/base.py (hunk 2 above)
Comment (antalszava) [reply pennylane/qnodes/base.py]: I double-checked this on a very simple example I had a look at and realized that `timeit.timeit` does not return results averaged over the number of runs. What this means, is that this result was actually in the regime of microseconds (a mistake I made, unfortunately).

This is the small example I looked at (now using the `%timeit` syntax which *does* return valid statistics):
```python
import inspect
def func(a, b, c, d=1234):
    return 2

%timeit inspect.getfullargspec(func)
```

Comment (antalszava) [reply pennylane/qnodes/base.py]: So the question would rather be if the flattening part would really introduce overhead. If not, then the logic in this PR should be simplified to removing caching only.

Comment (johannesjmeyer) [reply pennylane/qnodes/base.py]: If `functools.inspect` is the bottleneck, then we can also just cache the result of it, because this will not be subject to change even if the contents of the variables change

Comment (antalszava) [reply pennylane/qnodes/base.py]: Could it perhaps change if we have `**kwargs` in the signature of the qfunc? For example:
```python
@qml.qnode(dev)  
def circuit(a, **kwargs):  
    c = kwargs.get('c', False) 
    qml.RX(a, wires=[0]) 
    if c: 
        qml.RY(a, wires=[0]) 
    return qml.expval(qml.PauliZ(0))  
  
circuit(0.3) 
circuit(0.5, c=True) 
```

Comment (antalszava) [reply pennylane/qnodes/base.py]: PennyLane 0.7 results:

30 evaluations (`num_evaluations = 30`):

```python
In [2]: qml.about()                                                                                                                                                                                         
WARNING: pip is being invoked by an old script wrapper. This will fail in a future version of pip.
Please see https://github.com/pypa/pip/issues/5599 for advice on fixing the underlying issue.
To avoid this problem you can invoke Python with '-m pip' instead of running pip directly.
Name: PennyLane
Version: 0.7.0
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/XanaduAI/pennylane
Author: None
Author-email: None
License: Apache License 2.0
Location: /anaconda3/lib/python3.7/site-packages
Requires: scipy, autograd, toml, semantic-version, numpy, appdirs, networkx
Required-by: PennyLane-SF, PennyLane-Forest, PennyLane-Qchem, PennyLane-PQ, PennyLane-qsharp, PennyLane-Cirq
Platform info:           Linux-4.19.11-041911-generic-x86_64-with-debian-buster-sid
Python version:          3.7.6
Numpy version:           1.16.4
Scipy version:           1.4.1
Installed devices:
- default.gaussian (PennyLane-0.7.0)
- default.qubit (PennyLane-0.7.0)
- expt.tensornet (PennyLane-0.7.0)
- expt.tensornet.tf (PennyLane-0.7.0)
- strawberryfields.fock (PennyLane-SF-0.6.0)
- strawberryfields.gaussian (PennyLane-SF-0.6.0)
- forest.numpy_wavefunction (PennyLane-Forest-0.6.0)
- forest.qpu (PennyLane-Forest-0.6.0)
- forest.qvm (PennyLane-Forest-0.6.0)
- forest.wavefunction (PennyLane-Forest-0.6.0)
- projectq.classical (PennyLane-PQ-0.7.0.dev0)
- projectq.ibm (PennyLane-PQ-0.7.0.dev0)
- projectq.simulator (PennyLane-PQ-0.7.0.dev0)
- microsoft.QuantumSimulator (PennyLane-qsharp-0.6.0)
- cirq.simulator (PennyLane-Cirq-0.8.0)

In [5]: import pennylane as qml  
   ...: import numpy as np  
   ...:   
   ...: dev = qml.device('default.qubit', wires=1)  
   ...:   
   ...: @qml.qnode(dev)  
   ...: def circuit(a, b=1):  
   ...:     for idx in range(b):  
   ...:         qml.RX(a[idx], wires=[0])  
   ...:     return qml.expval(qml.PauliX(0))  
   ...:   
   ...: def func():  
   ...:     num_evaluations = 30  
   ...:     for i in range(num_evaluations):  
   ...:         params = np.random.rand(i)  
   ...:         circuit(params, b=i)  
   ...: %timeit func()                                                                                                                                                                                      
4.32 ms ± 34.2 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

100 evaluations (`num_evaluations = 100`):

```python
In [6]: import pennylane as qml  
   ...: import numpy as np  
   ...:   
   ...: dev = qml.device('default.qubit', wires=1)  
   ...:   
   ...: @qml.qnode(dev)  
   ...: def circuit(a, b=1):  
   ...:     for idx in range(b):  
   ...:         qml.RX(a[idx], wires=[0])  
   ...:     return qml.expval(qml.PauliX(0))  
   ...:   
   ...: def func():  
   ...:     num_evaluations = 100 
   ...:     for i in range(num_evaluations):  
   ...:         params = np.random.rand(i)  
   ...:         circuit(params, b=i)  
   ...: %timeit func()                                                                                                                                                                                      
27.2 ms ± 3.43 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

Here going from 30 to 100 evaluations yielded 6.2962962 times of increase in execution time.

--------------------------------

PennyLane 0.9.0-dev with no caching (simply removing the caching condition)

```python
In [10]: qml.about()                                                                                                                                                                                        
WARNING: pip is being invoked by an old script wrapper. This will fail in a future version of pip.
Please see https://github.com/pypa/pip/issues/5599 for advice on fixing the underlying issue.
To avoid this problem you can invoke Python with '-m pip' instead of running pip directly.
Name: PennyLane
Version: 0.9.0.dev0
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/XanaduAI/pennylane
Author: None
Author-email: None
License: Apache License 2.0
Location: /xanadu/pennylane
Requires: numpy, scipy, networkx, autograd, toml, appdirs, semantic-version
Required-by: PennyLane-SF, PennyLane-Forest, PennyLane-Qchem, PennyLane-PQ, PennyLane-qsharp, PennyLane-Cirq
Platform info:           Linux-4.19.11-041911-generic-x86_64-with-debian-buster-sid
Python version:          3.7.6
Numpy version:           1.16.4
Scipy version:           1.4.1
Installed devices:
- default.gaussian (PennyLane-0.9.0.dev0)
- default.qubit (PennyLane-0.9.0.dev0)
- default.tensor (PennyLane-0.9.0.dev0)
- default.tensor.tf (PennyLane-0.9.0.dev0)
- strawberryfields.fock (PennyLane-SF-0.6.0)
- strawberryfields.gaussian (PennyLane-SF-0.6.0)
- forest.numpy_wavefunction (PennyLane-Forest-0.6.0)
- forest.qpu (PennyLane-Forest-0.6.0)
- forest.qvm (PennyLane-Forest-0.6.0)
- forest.wavefunction (PennyLane-Forest-0.6.0)
- projectq.classical (PennyLane-PQ-0.7.0.dev0)
- projectq.ibm (PennyLane-PQ-0.7.0.dev0)
- projectq.simulator (PennyLane-PQ-0.7.0.dev0)
- microsoft.QuantumSimulator (PennyLane-qsharp-0.6.0)
- cirq.simulator (PennyLane-Cirq-0.8.0)

In [11]: import pennylane as qml  
    ...: import numpy as np  
    ...:   
    ...: dev = qml.device('default.qubit', wires=1)  
    ...:   
    ...: @qml.qnode(dev)  
    ...: def circuit(a, b=1):  
    ...:     for idx in range(b):  
    ...:         qml.RX(a[idx], wires=[0])  
    ...:     return qml.expval(qml.PauliX(0))  
    ...:   
    ...: def func():  
    ...:     num_evaluations = 30  
    ...:     for i in range(num_evaluations):  
    ...:         params = np.random.rand(i)  
    ...:         circuit(params, b=i)  
    ...: %timeit func()                                                                                                                                                                                     
90.7 ms ± 381 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

100 evaluations (`num_evaluations = 100`):

```python
In [12]: import pennylane as qml  
    ...: import numpy as np  
    ...:   
    ...: dev = qml.device('default.qubit', wires=1)  
    ...:   
    ...: @qml.qnode(dev)  
    ...: def circuit(a, b=1):  
    ...:     for idx in range(b):  
    ...:         qml.RX(a[idx], wires=[0])  
    ...:     return qml.expval(qml.PauliX(0))  
    ...:   
    ...: def func():  
    ...:     num_evaluations = 100 
    ...:     for i in range(num_evaluations):  
    ...:         params = np.random.rand(i)  
    ...:         circuit(params, b=i)  
    ...: %timeit func()                                                                                                                                                                                     
2.43 s ± 27.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
Here going from 30 to 100 evaluations yielded 26.791620728 times of increase in execution time.

--------------------------------

PennyLane 0.9.0-dev with cached `inspect.getfullargspec` (it is called only once per mutable QNode and the first result is used each time)
```python
In [1]: import pennylane as qml  
   ...: import numpy as np  
   ...:   
   ...: dev = qml.device('default.qubit', wires=1)  
   ...:   
   ...: @qml.qnode(dev)  
   ...: def circuit(a, b=1):  
   ...:     for idx in range(b):  
   ...:         qml.RX(a[idx], wires=[0])  
   ...:     return qml.expval(qml.PauliX(0))  
   ...:   
   ...: def func():  
   ...:     num_evaluations = 100 
   ...:     for i in range(num_evaluations):  
   ...:         params = np.random.rand(i)  
   ...:         circuit(params, b=i)  
   ...: %timeit func()                                                                                                                                                                                      
1.81 s ± 178 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

Comment (josh146) [reply pennylane/qnodes/base.py]: Thanks @antalszava! So, summing up:

* v0.9 with cached `inspect.getfullargspec`, 100 evals: 1.81 s ± 178 ms

* v0.9 with no caching, 100 evals: 2.43 s ± 27.3 ms

* v0.7 logic, 100 evals: 27.2 ms ± 3.43 ms

This to me is a strong argument that we should revert to v0.7 logic ASAP, and the continue work in a new PR to efficiently return the variable naming feature.

Comment (antalszava) [reply pennylane/qnodes/base.py]: Great, thanks for that! :blush:  (sorry for the long and cryptic logs :sweat_smile:)

File: pennylane/circuit_drawer/representation_resolver.py [hunk 8]
```diff
@@ -29,9 +29,12 @@ class RepresentationResolver:
...
+    def __init__(self, charset=UnicodeCharSet):
         self.charset = charset
-        self.show_variable_names = show_variable_names
+
+        # TODO: re-add the following attribute, once the string representation of VariableRefs
+        # is re-enabled
+        # self.show_variable_names = show_variable_names
```
Comment (smite) [inline pennylane/circuit_drawer/representation_resolver.py]: I'm with Josh here, wouldn't it be easier if you just made sure that `show_variable_names` is `False`? That way it's easier to put back on later.

**Comment by antalszava** [comment]
After some further exploring, it seems that the degradation in the runtime of the benchmarks was coming from [an older commit](https://github.com/XanaduAI/pennylane/commit/493d41995e1174761ce8dcffec0d292327a67310):

Reverting to the Copy button commit:
```
git reset --hard 3862206ff513ad2fe6f156174bb24ec9d28073a6
```
18.9 ms ± 18.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

--------------------------------------------------------------------

Reverting to the QNode deprecation commit:
```
git reset --hard 493d41995e1174761ce8dcffec0d292327a67310
```
1.49 s ± 2.31 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

--------------------------------------------------------------------
At this point, it seemed better to simply just go with the removal of caching (as suggested by @johannesjmeyer) with the following two possible paths to take for regaining the speed we experienced:
* Explore the pre-deprecate QNode logic (more promising)
* Meaningfully cache the value of `inspect.getfullargspec`

Thanks so much for the comments and ideas! :blush:

File: tests/qnodes/test_qnode_base.py [hunk 9]
```diff
@@ -830,6 +830,54 @@ def non_mutable_circuit(x, *, c=None):
...
+        res = circuit([phi, theta], n_layers=2)
+        exp = np.cos(phi + theta)
+        assert np.allclose(res, exp, atol=tol, rtol=0)
+
+        res = circuit([phi, theta, varphi], n_layers=3)
+        exp = np.cos(phi + theta + varphi)
+        assert np.allclose(res, exp, atol=tol, rtol=0)
+
+    def test_mutable_qnode_for_loop_varying_executions(self, tol):
+        """
```
Comment (josh146) [inline tests/qnodes/test_qnode_base.py]:
```suggestion
        """Test that a mutable QNode containing a for loop correctly mutates
        when called with different auxiliary arguments and different shaped positional
        arguments.
```
